Build fails for STM32/ARMv7 target

## Version(s) of meta-rust

Current master, 81d0fbfc023e987068ea1c035e33d6f0ad175cd1

Using Rust 1.78.

## Version(s) of poky and/or oe-core

kirkstone-4.0.17, commit 6d1a878bbf24c66f7186b270f823fcdf82e35383

## Expected result

The `rust-hello-world` recipe builds successfully

## Actual result

Trying to build `rust-hello-world` does not actually reach trying to build that recipe. Instead, building `libstd-rs` fails.
Abbreviated error output:

```
|      Running `rustc --crate-name libc --edition=2015 /workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/rustc-1.78.0-src/vendor/libc/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="align"' --cfg 'feature="rustc-dep-of-std"' --cfg 'feature="rustc-std-workspace-core"' -C metadata=ffe5355565292cd6 -C extra-filename=-ffe5355565292cd6 --out-dir /workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/build/arm-poky-linux-gnueabi/release/deps --target arm-poky-linux-gnueabi -C linker=/workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/wrapper/target-rust-ccld -L dependency=/workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/build/arm-poky-linux-gnueabi/release/deps -L dependency=/workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/build/release/deps --extern rustc_std_workspace_core=/workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/build/arm-poky-linux-gnueabi/release/deps/librustc_std_workspace_core-7df3a077613c532b.rmeta --cap-lints allow -L /workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/recipe-sysroot/usr/lib/rust --remap-path-prefix=/workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0=/usr/src/debug/libstd-rs/1.78.0-r0 -Cembed-bitcode=yes -L /workdir/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/libstd-rs/1.78.0-r0/recipe-sysroot/usr/lib -C link-arg=-Wl,-soname,libstd.so --cfg freebsd11 --cfg libc_priv_mod_use --cfg libc_union --cfg libc_const_size_of --cfg libc_align --cfg libc_int128 --cfg libc_core_cvoid --cfg libc_packedN --cfg libc_cfg_target_vendor --cfg libc_non_exhaustive --cfg libc_long_array --cfg libc_ptr_addr_of --cfg libc_underscore_const_names --cfg libc_thread_local --cfg libc_const_extern_fn`
| error: data-layout for target `arm-poky-linux-gnueabi`, `e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64`, differs from LLVM target's `armv7-unknown-linux-gnueabihf` default layout, `e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64`
```

Basically the data-layout is different between Rust's and LLVM's, I guess...?

## Steps to reproduce

I'm targeting a machine that uses an STM32MP1. Board support, etc. is from here: https://github.com/karo-electronics/meta-karo
I guess try to build `rust-hello-world` with current `meta-rust` for a machine that uses the armv7-eabi data layout...

## Additional information

Building the `rust-hello-world` recipe works fine with my Poky's version of Rust, and that, curiously, has the same `DATA_LAYOUT` set as `meta-rust`. I thought maybe the difference in version of `rust-llvm` might be the issue, but as far as I can tell, both versions of `rust-llvm` use the same data layout, too. I tried simply pasting the LLVM data layout from the error message into `rust-common.inc`, but to no avail. I did a `bitbake -c cleanall rust-native rust-llvm-native`, but that did not help, either..

I did notice b93853cb12e5e9cacb76c439d36eb6b96415026d which fixes basically the same issue but for x86.

Any help is appreciated, thanks!

## Update

I should have mentioned that all this happened while trying to build with Rust 1.78.0. I've added that information above.

I tried using Rust 1.75.0, and that fails at a different stage:

```
|      Running `/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/bootstrap/debug/rustc --crate-name rustc_baked_icu_data --edition=2021 compiler/rustc_baked_icu_data/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -Zunstable-options --check-cfg 'values(feature, "rustc_use_parallel_compiler")' --check-cfg 'names()' --check-cfg 'values()' -C metadata=d80321c1703c5ddc -C extra-filename=-d80321c1703c5ddc --out-dir /workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps --target x86_64-linux -C linker=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/wrapper/target-rust-cc -L dependency=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps -L dependency=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/release/deps --extern icu_list=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps/libicu_list-cda5c769a969b9c4.rmeta --extern icu_locid=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps/libicu_locid-a0307f8999234a04.rmeta --extern icu_locid_transform=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps/libicu_locid_transform-861103e127456362.rmeta --extern icu_provider=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps/libicu_provider-6a3919b9e9a2d926.rmeta --extern icu_provider_adapters=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps/libicu_provider_adapters-c0d119e935491acb.rmeta --extern zerovec=/workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-linux/release/deps/libzerovec-f28d2624ac7f349e.rmeta --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' '--check-cfg=values(emulate_second_only_system)' '--check-cfg=values(windows_raw_dylib)' -Zdual-proc-macros -Zmacro-backtrace -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Csplit-debuginfo=off -Zunstable-options '-Wrustc::internal' -Cprefer-dynamic -Z binary-dep-depinfo`
| error[E0308]: mismatched types
|     --> /workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/compiler/rustc_index/src/lib.rs:47:32
|      |
| 45   | macro_rules! static_assert_size {
|      | ------------------------------- in this expansion of `static_assert_size!`
| 46   |     ($ty:ty, $size:expr) => {
| 47   |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
|      |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 24 elements, found one with 32 elements
|      |
|     ::: compiler/rustc_ast/src/ast.rs:3163:34
|      |
| 3163 |     static_assert_size!(LitKind, 24);
|      |     --------------------------------
|      |     |                            |
|      |     |                            help: consider specifying the actual array length: `32`
|      |     in this macro invocation
|
| error[E0308]: mismatched types
|     --> /workdir/build/tmp/work/x86_64-linux/rust-native/1.75.0-r0/rustc-1.75.0-src/compiler/rustc_index/src/lib.rs:47:32
|      |
| 45   | macro_rules! static_assert_size {
|      | ------------------------------- in this expansion of `static_assert_size!`
| 46   |     ($ty:ty, $size:expr) => {
| 47   |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
|      |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 40 elements, found one with 48 elements
|      |
|     ::: compiler/rustc_ast/src/ast.rs:3165:38
|      |
| 3165 |     static_assert_size!(MetaItemLit, 40);
|      |     ------------------------------------
|      |     |                                |
|      |     |                                help: consider specifying the actual array length: `48`
|      |     in this macro invocation
|
```

The above error occurs while trying to build rust-native.

--- a/recipes-devtools/rust/rust-common.inc
+++ b/recipes-devtools/rust/rust-common.inc
@@ -121,7 +121,7 @@ def llvm_features(d):
 
 
 ## arm-unknown-linux-gnueabihf
-DATA_LAYOUT[arm-eabi] = "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
+DATA_LAYOUT[arm-eabi] = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
 LLVM_TARGET[arm-eabi] = "${RUST_TARGET_SYS}"
 TARGET_ENDIAN[arm-eabi] = "little"
 TARGET_POINTER_WIDTH[arm-eabi] = "32"
@@ -130,7 +130,7 @@ MAX_ATOMIC_WIDTH[arm-eabi] = "64"
 FEATURES[arm-eabi] = "+v6,+vfp2"
 
 ## armv7-unknown-linux-gnueabihf
-DATA_LAYOUT[armv7-eabi] = "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
+DATA_LAYOUT[armv7-eabi] = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
 LLVM_TARGET[armv7-eabi] = "${RUST_TARGET_SYS}"
 TARGET_ENDIAN[armv7-eabi] = "little"
 TARGET_POINTER_WIDTH[armv7-eabi] = "32"

--- a/recipes-devtools/rust/rust-common.inc
+++ b/recipes-devtools/rust/rust-common.inc
@@ -378,4 +378,11 @@ python do_rust_gen_targets () {
 
 addtask rust_gen_targets after do_patch before do_compile
 do_rust_gen_targets[dirs] += "${WORKDIR}/targets"
-
+do_rust_gen_targets[vardeps] += "\
+    LLVM_TARGET \
+    DATA_LAYOUT \
+    MAX_ATOMIC_WIDTH \
+    TARGET_POINTER_WIDTH \
+    TARGET_C_INT_WIDTH \
+    TARGET_ENDIAN \
+"

